cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java

```diff
@@ -0,0 +1,302 @@
+   * service.
+   * One callback is created per TopicPartition. It is created by the first subscription to a feed
+   * which maps to the TopicPartition.
+   */
+  private final class KafkaNotificationsCallback implements KafkaConsumer.MessageCallback, Cancellable {
+
+    private final Multimap<NotificationFeed, Cancellable> subscribers;
+    private final Cancellable kafkSubscription;
+    private final TopicPartition topicPartition;
+
+    private KafkaNotificationsCallback(TopicPartition topicPartition) {
+      this.topicPartition = topicPartition;
+      this.subscribers = Multimaps.synchronizedMultimap(HashMultimap.<NotificationFeed, Cancellable>create());
+
+      KafkaConsumer.Preparer preparer = kafkaClient.getConsumer().prepare();
+
+      // TODO there is a bug in twill, that when the topic doesn't exist, add latest will not make subscription
+      // start from offset 0 - but that will be fixed soon
+      preparer.addLatest(topicPartition.getTopic(), topicPartition.getPartition());
+      kafkSubscription = preparer.consume(this);
+    }
+
+    @Override
+    public void onReceived(Iterator<FetchedMessage> messages) {
+      int count = 0;
+      while (messages.hasNext()) {
+        count++;
+        FetchedMessage message = messages.next();
+        ByteBuffer payload = message.getPayload();
+
+        try {
+          KafkaMessage decodedMessage = KafkaMessageCodec.decode(payload);
+          try {
+            delegate.publish(KafkaNotificationUtils.getMessageFeed(decodedMessage.getMessageKey()),
+                             decodedMessage.getNotificationJson());
+          } catch (Throwable t) {
+            LOG.warn("Error while processing notification {} with handler {}", decodedMessage.getNotificationJson(), t);
+          }
+        } catch (IOException e) {
+          LOG.error("Could not decode Kafka message {} using Gson.", message, e);
+        }
+      }
+      LOG.debug("Handled {} messages from kafka", count);
+    }
+
+    /**
+     * Subscribe to a {@code feed} with a {@code handler}, using the delegate {@link InMemoryNotificationService}.
+     *
+     * @param feed {@link NotificationFeed} to subscribe to
+     * @param handler {@link NotificationHandler} to use to process received notifications
+     * @param executor {@link Executor} used to push notifications to the handler
+     * @param <N> Type of the notifications to handle
+     * @return a {@link Cancellable} to cancel the subscription
+     * @throws NotificationFeedException if the subscription throught the {@link InMemoryNotificationService} could
+     * not be made
+     */
+    public <N> Cancellable addSubscription(final NotificationFeed feed, NotificationHandler<N> handler,
+                                           Executor executor) throws NotificationFeedException {
+      final Cancellable cancellable = delegate.subscribe(feed, handler, executor);
+      subscribers.put(feed, cancellable);
```

Comment:
This class still needs to remember a few things to know to which Kafka topic partition it has to listen to, and which ones it should stop listening to. This is the purpose of the ``subscribers`` list here.